connexion/apps/asynchronous.py

```diff
@@ -0,0 +1,123 @@
+from connexion.middleware.abstract import (
+    AbstractRoutingAPI,
+    RoutedAPI,
+    RoutedMiddleware,
+)
+from connexion.operations import AbstractOperation
+from connexion.uri_parsing import AbstractURIParser
+
+logger = logging.getLogger("Connexion.app")
+
+
+class AsyncOperation:
+    def __init__(
+        self,
+        operation: AbstractOperation,
+        fn: t.Callable,
+        uri_parser: AbstractURIParser,
+        api: AbstractRoutingAPI,
+        operation_id: str,
+        pythonic_params: bool,
+    ) -> None:
+        self._operation = operation
+        self._fn = fn
+        self.uri_parser = uri_parser
+        self.api = api
+        self.operation_id = operation_id
+        self.pythonic_params = pythonic_params
+
+    @classmethod
+    def from_operation(
+        cls, operation: AbstractOperation, pythonic_params: bool
+    ) -> "AsyncOperation":
+        return cls(
+            operation,
+            fn=operation.function,
+            uri_parser=operation.uri_parser_class,
+            api=operation.api,
+            operation_id=operation.operation_id,
+            pythonic_params=pythonic_params,
+        )
+
+    @property
+    def fn(self) -> t.Callable:
+        decorator = StarletteDecorator(
+            pythonic_params=self.pythonic_params,
+            jsonifier=self.api.jsonifier,
+        )
+        return decorator(self._fn)
+
+    async def __call__(
+        self, scope: Scope, receive: Receive, send: Send
+    ) -> StarletteResponse:
+        return await self.fn(scope=scope, receive=receive, send=send)
+
+
+class AsyncApi(RoutedAPI[AsyncOperation]):
+    def __init__(self, *args, pythonic_params: bool, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.pythonic_params = pythonic_params
+        self.router = Router()
+        self.add_paths()
+
+    def make_operation(self, operation: AbstractOperation) -> AsyncOperation:
+        return AsyncOperation.from_operation(operation, self.pythonic_params)
+
+
+class AsyncMiddlewareApp(RoutedMiddleware[AsyncApi]):
+    def __init__(self) -> None:
+        self.apis: t.Dict[str, AsyncApi] = {}
+        self.operations: t.Dict[str, AsyncOperation] = {}
+        self.router = Router()
+        super().__init__(self.router)
+
+    def add_api(self, *args, **kwargs):
+        api = AsyncApi(*args, **kwargs)
+        self.apis[api.base_path] = api
+        self.router.mount(api.base_path, api.router)
+        return api
+
+    def add_url_rule(
+        self,
+        rule,
+        endpoint: str = None,
+        view_func: t.Callable = None,
+        methods: t.List[str] = None,
+        **options
+    ):
+        self.router.add_route(rule, endpoint=view_func, name=endpoint, methods=methods)
+
+
+class AsyncApp(AbstractApp):
+    """Connexion Application based on ConnexionMiddleware wrapping a async Connexion application
+    based on starlette tools."""
+
+    middleware_app = AsyncMiddlewareApp()
```

Comment:
Is it intentional that this is a class attribute and not an instance attribute?

Reply:
Not really, I just didn't want to repeat the whole `__init__` method with all its parameters here. But I guess this could prove to be a problem with parallelism / concurrency. Will change it.